### PR TITLE
[SG-65] Reusable Registration Form

### PR DESCRIPTION
**Pull request by gbubemismith**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Create a reusable registration form that can be used in the trail initiation vertical stepper. I increased the scope a bit by changing the form to a reactive form and updating this changes on the other clients.

## Code changes

- **libs/angular/src/components/register.component.ts:** Switched from template driven forms to reactive forms.
- **apps/web/src/app/modules/register-form/register-form.module.ts:** Created register form module so component can be shared easily.
- **apps/web/src/app/modules/register-form/register-form.component.ts:** Register form component class which relies on `BaseRegisterComponent` for the reactive form creation, validation and submission of the form.
- **apps/web/src/app/modules/register-form/register-form.component.html:** Register form template which uses tailwind classes and the new Bitwarden components.
- **apps/web/src/app/accounts/register.component.html:** Replaced current form with the reusable `register-form`
- **apps/desktop/src/app/accounts/register.component.html:** Converted the desktop client registration form to reactive forms.
- **apps/browser/src/popup/accounts/register.component.html:** Converted the browser extension client registration form to reactive forms.

## Screenshots

https://user-images.githubusercontent.com/13024008/174962766-768e7ad0-f7bb-437b-870e-f208bdbc09f4.mov

## Before you submit

```
- [X] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

### CODE REVIEW

File: apps/web/src/app/modules/register-form/register-form.component.html [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            [ngClass]="{ 'bwi-eye': !showPassword, 'bwi-eye-slash': showPassword }"
+          ></i>
+        </button>
+        <bit-hint class="tw-text-sm tw-mt-2.5">
+          <span class="tw-font-semibold">Important:</span>
+          {{ "masterPassImportant" | i18n }}</bit-hint
+        >
+      </bit-form-field>
+      <app-password-strength [score]="masterPasswordScore" [showText]="true">
+      </app-password-strength>
```
[the remark this answers is not in the document]
Comment (gbubemismith) [reply apps/web/src/app/modules/register-form/register-form.component.html:53]: @Hinton thanks for this. A separate ticket has been created for this and would be worked on.

File: libs/angular/src/components/register.component.ts [hunk 2]
```diff
@@ -266,4 +228,62 @@ export class RegisterComponent extends CaptchaProtectedComponent implements OnIn
...
+              hintEqualsPassword: {
+                message: this.i18nService.t("hintEqualsPassword"),
+              },
+            }
+          : null;
+      }
+
+      return null;
+    };
+  }
```
Comment (Hinton) [inline libs/angular/src/components/register.component.ts:262]: I would suggest making a generic rule for this. That accepts the other field, and the error message to display.

We should be able to just add another variable to the function and run it through the `i18nService`.

File: libs/common/src/services/formValidationErrors.service.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            controlName: key,
+            errorName: keyError,
+          });
+        });
+      }
+    });
+
+    return errors;
+  }
+}
```
Comment (differsthecat) [inline libs/common/src/services/formValidationErrors.service.ts:31]: Is the purpose of this to show multiple validation errors? Was there not a way to show multiple validation errors without this service?

Comment (gbubemismith) [reply libs/common/src/services/formValidationErrors.service.ts:31]: This helps me get all validation errors on the form which is then used to display the toast for the other clients in dynamic way

File: apps/web/src/app/modules/register-form/register-form.component.html [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        class="tw-px-6 tw-py-2.5"
+        type="submit"
+        bitButton
+        buttonType="primary"
+        class="btn-submit"
+        [disabled]="form.loading"
+      >
+        <span>{{ "createAccount" | i18n }}</span>
+        <i class="bwi bwi-spinner bwi-spin" title="{{ 'loading' | i18n }}" aria-hidden="true"></i>
+      </button>
```
[the remark this answers is not in the document]
Comment (Hinton) [reply apps/web/src/app/modules/register-form/register-form.component.html]: https://github.com/bitwarden/clients/pull/2971

**Pull request merged**